### PR TITLE
Fluentd cloudwatch

### LINKED ISSUE #487: operator support for fluentd kinesis/cloudwatch plugins
### Is your feature request related to a problem? Please describe.

I have fluent operator enhanced to support using the fluentd kinesis plugin and am working on cloudwatch plugin support.  Is there interest in having this added to the main codebase, or was there a conscious decision to limit the fluentd plugins that are supported?

### Describe the solution you'd like

.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jseutter**
### What this PR does / why we need it:
This PR supercedes https://github.com/fluent/fluent-operator/pull/586 by rebasing my changes against master.

This change adds support for the cloudwatch plugin in fluentd which is able to send logs from Fluentd to AWS Cloudwatch.

I tested this change in OpenShift, not sure if that is significant or not.

### Which issue(s) this PR fixes:
Fixes #487 

### Does this PR introduced a user-facing change?

```release-note
Added support for the fluentd cloudwatch plugin to send logs to AWS Cloudwatch
```

### Additional documentation, usage docs, etc.:

```docs

```

### CODE REVIEW

File: VERSION [hunk 1]
```diff
@@ -1 +1 @@
-latest
+v1.7.0
```
Comment (benjaminhuo) [inline VERSION:1]: The version should be latest instead of v1.7.0 for master branch

File: apis/fluentd/v1alpha1/plugins/output/types.go [hunk 2]
```diff
@@ -46,6 +47,8 @@ type Output struct {
 	Loki *Loki `json:"loki,omitempty"`
 	// Custom plugin type
 	CustomPlugin *custom.CustomPlugin `json:"customPlugin,omitempty"`
+	// out_cloudwatch plugin
+	Cloudwatch *Cloudwatch `json:"cloudwatch,omitempty"`
```
Comment (benjaminhuo) [inline apis/fluentd/v1alpha1/plugins/output/types.go:51]:
```suggestion
	CloudWatch *CloudWatch `json:"cloudWatch,omitempty"`
```

File: apis/fluentd/v1alpha1/plugins/output/cloudwatch.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
+package output
+
+// Cloudwatch defines the parametes for out_cloudwatch output plugin
+type Cloudwatch struct {
```
Comment (benjaminhuo) [inline apis/fluentd/v1alpha1/plugins/output/cloudwatch.go:4]: I think we should change `Cloudwatch` to `CloudWatch` following the official name:
image

File: apis/fluentd/v1alpha1/plugins/output/types.go [hunk 4]
```diff
@@ -596,6 +604,155 @@ func (o *Output) lokiPlugin(parent *params.PluginStore, loader plugins.SecretLoa
 	return parent
 }
 
+func (o *Output) cloudwatchPlugin(parent *params.PluginStore, sl plugins.SecretLoader) *params.PluginStore {
```
Comment (benjaminhuo) [inline apis/fluentd/v1alpha1/plugins/output/types.go:607]: cloudWatchPlugin

File: apis/fluentd/v1alpha1/plugins/params/const.go [hunk 5]
```diff
@@ -34,6 +34,7 @@ const (
 	OpensearchPlugin    PluginName = "opensearch"
 	MatchPlugin         PluginName = "match"
 	BufferPlugin        PluginName = "buffer"
+	CloudwatchPlugin    PluginName = "cloudwatch_logs"
```
Comment (benjaminhuo) [inline apis/fluentd/v1alpha1/plugins/params/const.go:37]: CloudWatchPlugin

File: apis/fluentd/v1alpha1/tests/tools.go [hunk 6]
```diff
@@ -308,14 +308,31 @@ metadata:
...
+	FluentdClusterOutput2CloudwatchRaw = `
+apiVersion: fluentd.fluent.io/v1alpha1
+kind: ClusterOutput
+metadata:
+  name: fluentd-output-cloudwatch
+  labels:
+    output.fluentd.fluent.io/enabled: "true"
+spec:
+  outputs:
+  - cloudwatch:
```
Comment (benjaminhuo) [inline apis/fluentd/v1alpha1/tests/tools.go:329]: cloudWatch

File: charts/fluent-operator/values.yaml [hunk 7]
```diff
@@ -16,7 +16,7 @@ operator:
     tag: "20.10"
   container:
     repository: "kubesphere/fluent-operator"
-    tag: "latest"
+    tag: "v1.7.0"
```
Comment (benjaminhuo) [inline charts/fluent-operator/values.yaml:19]:
```suggestion
    tag: "latest"
```

File: manifests/setup/setup.yaml [hunk 8]
```diff
@@ -19233,7 +19441,7 @@ spec:
             fieldRef:
               apiVersion: v1
               fieldPath: metadata.namespace
-        image: kubesphere/fluent-operator:latest
+        image: kubesphere/fluent-operator:v1.7.0
```
Comment (benjaminhuo) [inline manifests/setup/setup.yaml:19444]:
```suggestion
        image: kubesphere/fluent-operator:latest
```

File: VERSION (hunk 1 above)
Comment (benjaminhuo) [inline VERSION:1]:
```suggestion
latest
```

File: manifests/setup/fluent-operator-deployment.yaml [hunk 9]
```diff
@@ -39,7 +39,7 @@ spec:
           mountPath: /var/run/docker.sock
       containers:
       - name: fluent-operator
-        image: kubesphere/fluent-operator:latest
+        image: kubesphere/fluent-operator:v1.7.0
```
Comment (benjaminhuo) [inline manifests/setup/fluent-operator-deployment.yaml:42]:
```suggestion
        image: kubesphere/fluent-operator:latest
```

File: apis/fluentd/v1alpha1/plugins/output/cloudwatch.go (hunk 3 above)
Comment (jseutter) [reply apis/fluentd/v1alpha1/plugins/output/cloudwatch.go:4]: Agreed.  I'll make the change.

File: apis/fluentd/v1alpha1/plugins/output/cloudwatch.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+	//
+	AutoCreateStream *bool `json:"autoCreateStream,omitempty"`
+	//
+	AwsKeyId *string `json:"awsKeyId,omitempty"`
```
[the remark this answers is not in the document]
Comment (jseutter) [reply apis/fluentd/v1alpha1/plugins/output/cloudwatch.go:8]: I'm struggling with this one a bit.  Does the fluent operator code copy the string from the secret into the fluentd config file?  That's the only way I can see this working without having to delve into the cloudwatch plugin code.

[the remark this answers is not in the document]
Comment (wenchajun) [reply apis/fluentd/v1alpha1/plugins/output/cloudwatch.go:8]: You can refer to this example for processing.
https://github.com/fluent/fluent-operator/blob/master/apis/fluentd/v1alpha1/plugins/output/es.go#L26
https://github.com/fluent/fluent-operator/blob/master/apis/fluentd/v1alpha1/plugins/output/types.go#L377

**Comment by benjaminhuo** [comment]
@jseutter would you help rebase your PR to resolve the conflict:
```shell
git remote add upstream https://github.com/fluent/fluent-operator.git
git fetch --all
git rebase upstream/master
```

**Comment by jseutter** [comment]
Changes made, branch rebased and tested.  I think this is ready for another look.

**Review by benjaminhuo — APPROVED** [review]

**Comment by benjaminhuo** [comment]
@jseutter That's great, thanks for the patience and efforts

**Pull request merged**